shipwright/source_control.py

```diff
@@ -52,6 +55,40 @@ def _hexsha(ref):
         return ref.hexsha[:12]
 
 
+def _hash_blob(blob):
+    if blob.hexsha != blob.NULL_HEX_SHA:
+        return blob.hexsha
+    else:
+        return blob.repo.git.hash_object(blob.abspath)
+
+
+def _dirty_suffix(repo, base_paths=['.']):
+    def abspath(path):
+        return os.path.abspath(os.path.join(repo.working_dir, path))
+
+    def hash_blobs(blobs):
+        return [(b.path, _hash_blob(b)) for b in blobs if b]
+
+    def in_paths(path):
+        return any(abspath(path).startswith(abspath(base_path) + os.sep)
+                   for base_path in base_paths)
+
+    diff = repo.head.commit.diff(None)
+    a_hashes = hash_blobs(map(lambda d: d.a_blob, diff))
+    b_hashes = hash_blobs(map(lambda d: d.b_blob, diff))
+    untracked_hashes = [(path, repo.git.hash_object(path))
+                        for path in repo.untracked_files]
+    hashes = sorted(a_hashes) + sorted(b_hashes + untracked_hashes)
+    filtered_hashes = [(path, h) for path, h in hashes if in_paths(path)]
+
+    if not filtered_hashes:
+        return ''
+    digest = hashlib.sha256()
+    for path, h in filtered_hashes:
+        digest.update(path.encode('utf-8') + b'\0' + h.encode('utf-8'))
+    return '-dirty-' + binascii.hexlify(digest.digest())[:12].decode('utf-8')
+
+
```

Comment:
I think this could be more performant with a function that takes a repo, and a list of all path dependencies:

```python
_all_dirty_suffix(repo, all_paths=[[pathA,], [pathA, pathB, ], [pathA, pathB, pathC]])
```

and is the equivalent of:
```python
def _all_dirty_suffix(repo, all_paths):
    [_dirty_suffix(repo, paths) for paths in all_paths)]
```

except it maintains only one untacked_hashes object and one hashes object